MaxText/layers/pipeline.py

```diff
@@ -217,18 +298,46 @@ def run_one_iteration(self, loop_state, positions, segment_ids, deterministic, m
    '''Run one loop iteration - gets weights and inputs for each stage, run the stages in parallel, and update the loop state.'''
    state_io = loop_state['state_io']
    shift = loop_state["shift"]
+   circ_storage = loop_state["circ_storage"]
    loop_iteration = loop_state["loop_iteration"]
 
-   microbatch_ids = self.get_microbatch_ids(loop_iteration)
+   microbatch_ids, _ = self.get_microbatch_and_repeat_ids(loop_iteration)
 
-   stages_inputs = self.get_iteration_inputs(loop_iteration, state_io, shift)
+   stages_inputs = self.get_iteration_inputs(loop_iteration, state_io, circ_storage, shift)
    # We checkpoint stages_inputs since we are grabbing only one slice of the state_io, don't need to save the entire buffer.
    stages_inputs = jax.ad_checkpoint.checkpoint_name(stages_inputs, 'iteration_input')
    stages_positions = self.vmap_gather(positions, microbatch_ids, 0) if positions is not None else None
    stages_segment_ids = self.vmap_gather(segment_ids, microbatch_ids, 0) if segment_ids is not None else None
 
    vmap_func = self.get_main_vmap_func()
 
+   if self.config.num_pipeline_repeats > 1:
+    _, repeat_ids = self.get_microbatch_and_repeat_ids(loop_iteration)
+
+    def prepare_vars_for_main_vmap(weights):
+      def gather_weights_for_stages_in(weights):
+        return jax.tree.map(
+            functools.partial(
+                self.vmap_parallel_gather, repeat_ids=repeat_ids, repeat_dim_in_weights=0, stages_dim_in_weights=1),
+            weights)
+      circular_metadata_params={
+        nn.PARTITION_NAME: "circular_repeats",
+        'sub_weight_split_dims_mapping': (None,),
+        "is_initializing": True,
```

Comment:
Should `is_initializing` always be True here?

Reply:
Hmm this is a good catch, replaced it with self.is_initializing(). Not entirely sure what behavior that changes other than denoting that the parameters can be modified